### PR TITLE
Add ability to not wait on child workflows in test suite.

### LINKED ISSUE #614: Testing cron child workflow with parent policy abandon runs forever
## Expected Behavior
Testing a cron child workflow with PARENT_CLOSE_POLICY_ABANDON should either close with the parent workflow for testsuite or have a method to shutdown the child workflow for the test.

## Actual Behavior
The cron child workflow runs forever due to the auto-fire timer and then panics when the test times out.

## Steps to Reproduce the Problem

Originally produced while updating go-sdk in a repo. There is a test below for internal_workflow_testsuite_test.go which can reproduce the issue
```
func (s *WorkflowTestSuiteUnitTest) Test_CronChildWorkflowWithParentPolicyAbandon() {
	failedCount, successCount, lastCompletionResult := 0, 0, 0
	cronWorkflow := func(ctx Context) (int, error) {
		info := GetWorkflowInfo(ctx)
		var result int
		if HasLastCompletionResult(ctx) {
			_ = GetLastCompletionResult(ctx, &result)
		}
		_ = Sleep(ctx, time.Second*3)
		if info.Attempt == 1 {
			failedCount++
			return 0, errors.New("please-retry")
		}

		successCount++
		result++
		lastCompletionResult = result
		return result, nil
	}

	testWorkflow := func(ctx Context) error {
		ctx1 := WithChildWorkflowOptions(ctx, ChildWorkflowOptions{
			WorkflowRunTimeout: time.Minute * 10,
			RetryPolicy: &RetryPolicy{
				MaximumAttempts:        5,
				InitialInterval:        time.Second,
				MaximumInterval:        time.Second * 10,
				BackoffCoefficient:     2,
				NonRetryableErrorTypes: []string{"bad-bug"},
			},
			ParentClosePolicy: enumspb.PARENT_CLOSE_POLICY_ABANDON,
			CronSchedule:      "0 * * * *", // hourly
		})

		cronFuture := ExecuteChildWorkflow(ctx1, cronWorkflow) // cron never stop so this future won't return

		timeoutTimer := NewTimer(ctx, time.Hour*3)
		selector := NewSelector(ctx)
		var err error
		selector.AddFuture(cronFuture, func(f Future) {
			err = errors.New("cron workflow returns, this is not expected")
		}).AddFuture(timeoutTimer, func(f Future) {
			// err will be nil
		}).Select(ctx)

		return err
	}

	env := s.NewTestWorkflowEnvironment()
	env.RegisterWorkflow(cronWorkflow)
	env.RegisterWorkflow(testWorkflow)

	startTime, _ := time.Parse(time.RFC3339, "2018-12-20T16:30:00+08:00")
	env.SetStartTime(startTime)
	env.ExecuteWorkflow(testWorkflow)

	s.True(env.IsWorkflowCompleted())
	err := env.GetWorkflowError()
	s.NoError(err)

	s.Equal(4, failedCount)
	s.Equal(4, successCount)
	s.Equal(4, lastCompletionResult)
}
```

Then you can run it with `go test -run "TestUnitTestSuite/Test_CronChildWorkflowWithParentPolicyAbandon"` in `sdk-go/internal`

## Specifications

  - Version: v1.11.0
  - Platform: Mac v11.6

### PR DESCRIPTION
**Pull request by cretz**
## What was changed

Added `TestWorkflowEnvironment.SetDetachedChildWait` that lets test environments not wait on detached (i.e. abandoned or request-cancel) children before finishing workflow.

## Why?

In #470, we introduced a change that causes a parent `ExecuteWorkflow` to not return until all of its children are completely done. This caused issues for people testing interminable child workflows such as ones started on cron because they never finish.

Alternatively, we could have done something like expose child test workflow environments so people could cancel children themselves. This seems more straightforward however.

## Checklist

1. Closes #614

### CODE REVIEW

File: internal/workflow_testsuite.go
Comment (bergundy) [inline internal/workflow_testsuite.go:545]: This sentence is unclear to me:
> regardless of detached child
Can you please clarify?

File: internal/workflow_testsuite.go
Comment (bergundy) [inline internal/workflow_testsuite.go:546]: Also the phrasing here is confusing, what does "allow" mean? Will it always wait for at least a single execution?

**Review by bergundy — COMMENTED** [review]
Is this the right approach?
Wouldn't an explicit shutdown call when all test assertions are fulfilled be a better experience?

File: internal/workflow_testsuite.go
Comment (cretz) [reply internal/workflow_testsuite.go:545]: It means, regardless of whether the child is detached or not, `ExecuteWorkflow` will wait for all to complete. This is the default behavior today and the problematic one.

**Comment by cretz** [comment]
>  Wouldn't an explicit shutdown call when all test assertions are fulfilled be a better experience?

The `ExecuteWorkflow` call never completes because there are child workflows that never complete. They used to complete before #470 was merged.

File: internal/workflow_testsuite.go
Comment (tszucs-stripe) [reply internal/workflow_testsuite.go:545]: It is a tricky thing to describe. Perhaps something like this might be clearer?
> If false, ExecuteWorkflow will block until all attached child workflows have completed and all detached child workflows have ran only once. This is useful when testing endless cron-based child workflows, as without it a detached child workflow would run forever.

**Review by tszucs-stripe — APPROVED** [review]
Not a code owner but this makes sense, I think enables users to test both situations, and does so in a simple manner with a bool flag.

File: internal/workflow_testsuite.go
Comment (cretz) [reply internal/workflow_testsuite.go]: Clarified

File: internal/workflow_testsuite.go
Comment (cretz) [reply internal/workflow_testsuite.go]: Clarified.

**Comment by cretz** [comment]
I have now learned there is no guarantee the child will even run, it depends on order of selector and other things. Going to make it clear in the docs.

File: internal/internal_workflow_testsuite.go
Comment (Sushisource) [inline internal/internal_workflow_testsuite.go:697]: Can this ever return true?

Comment (Sushisource) [reply internal/internal_workflow_testsuite.go:697]: Oh nevermind github just cut off the last line

**Review by Sushisource — APPROVED** [review]

**Pull request merged**
